### ISSUE TITLE
Type issues

### ISSUE THREAD

**Issue by RebeccaStevens** [vitaly-t/iter-ops#143]
I've looked into the type issues we are having, and it seems like we may be hitting a TypeScript bug. I've made an [issue here](https://github.com/microsoft/TypeScript/issues/51092).

I'll see if I can come up with a work-around.

**Comment by RebeccaStevens** [comment]
Looks like we'll have to avoid creating any intersections of functions.

**Comment by RebeccaStevens** [comment]
This definition seems to resolve most type issues:

```ts
type Operation<T, R> = (
    i: AsyncIterable<T> | Iterable<T>
) => AsyncIterable<R> & Iterable<R>
```

However, it isn't "correct" and thus if we go with it, we'd want to only use it internally (not export it). I'll see if I can come up with a better solution.

**Comment by RebeccaStevens** [comment]
I think the only way to properly get around this issue is to split all operations into explicit sync and async version. However, this will not be as elegant, so I understand not wanting to take this approach.

Originally, I was hoping that splitting the `pipe` function into a sync and async version would be enough, but unfortunately not. I would still however recommend keeping the split pipelines as it means users don't have to explicitly convert a non-async iterate before using the async pipeline (the `pipeSync` function can potentially be dropped).

**Comment by RebeccaStevens** [comment]
@vitaly-t I've done a bunch more looking into this and things aren't looking great 😕 

I've spent the last 8 hours or so converting all the function overloads to a single function definition that uses conditional types in hope that this would get around the TypeScript limitation/bug. This involved having to change the definition of `AsyncOperation` and `SyncOperation` as well as completely removing the main `Operation`

```ts
// New
type SyncOperation<
    T extends Iterable<unknown>,
    R extends Iterable<unknown>
> = (i: T) => R;

// Old
type SyncOperation<T, R> = (i: Iterable<T>) => Iterable<R>;
```

Although this did result in improvements with some operators, others had no improvements. I think the only way we can actually get the types working right is to split the operations.

Maybe we could do something like this:

```ts
import {pipe, filter, map} from 'iter-ops/sync';

const i = pipe(
    [1, 2, 3, 4, 5],
    filter((a) => a % 2 === 0),
    map((value) => ({value}))
);
```

```ts
import {pipe, distinct, delay} from 'iter-ops/async';

const i = pipe(
    [1, 2, 2, 3, 3, 4],
    distinct(),
    delay(1000)
);
```

This approach will allow for the sync and async operations to keep the same names and should allow for all the type issues to be solved.

I imagine that usually users will only want to use one of the pipelines at a time so this should be a non-issue. If users want to use both pipelines, they could possibly do something like this.

```ts
import {sync, async} from 'iter-ops';

const i = sync.pipe(
    [1, 2, 3, 4, 5],
    sync.filter((a) => a % 2 === 0),
    sync.map((value) => ({value}))
);

const j = async.pipe(
    [1, 2, 2, 3, 3, 4],
    async.distinct(),
    async.delay(1000)
);
```

Out of all the option we have, I think splitting the operations is the best solution. With every other solution I've come up with, it doesn't seem to be possible to get the types right.

I wish TypeScript would solve that issue, but as it seems to be a design limitation, the only way it will ever be solved is if TypeScript itself undergoes a major rewrite, which is unlikely to happen any time soon.

**Comment by RebeccaStevens** [comment]
Also, splitting the operations will allow for the sync and async versions to be tree-shaken from each other, thus this would be a performance optimization.

**Comment by vitaly-t** [comment]
Please remind me, what was the initial issue of changing so much code that's in the current v1.6.4 release? Did the types not work there correctly?

**Comment by RebeccaStevens** [comment]
> Please remind me, what was the initial issue of changing so much code that's in the current v1.6.4 release? Did the types not work there correctly?

The types "work" there but not in all cases. These cases are what I'm trying to fix. Fixing them should also allow for all of the `any`s and casts to be removed from the code base. Currently they can't be as TypeScript will then detect the cases where they don't work and complain.

I plan on adding a bunch more type test to test the types. The current type tests are just "happy path" tests.

> But that's a duplicate, for which a fix was released awhile ago?

It's a duplicate of https://github.com/microsoft/TypeScript/issues/26591 but there is no fix. That issue was labeled with "Design Limitation" which means "No apparent way of fixing this exists (short of wholescale redesign)". So TypeScript wants to fix the issue but they can't.

> P.S. I kind of really wish now that all this was a development branch instead... I might in fact force-split it now, because the state in which the master now isn't great. I cannot see at this point when the next update will be possible to publish.

I've reverted a bunch of the changes that were making the issues more apparent. This should make things look cleaner now like they did pre the 2.0 betas (though the underlying issues still remain).

**Comment by vitaly-t** [comment]
I tried everything in the updated master branch, including deleting all temporary folders and reinstalling, but tests still won't run anymore:

![image](https://user-images.githubusercontent.com/5108906/194780027-1699ffce-cc51-436f-96ba-a2045e30e549.png)

There's some issue with the `rollup` library. I'm investigating...

**Comment by vitaly-t** [comment]
`npm run doc` in `master` also doesn't work anymore:

![image](https://user-images.githubusercontent.com/5108906/194780137-c8d08483-0897-4441-992f-859fab02e977.png)

**Comment by vitaly-t** [comment]
After looking a little closer into issues with running tests....

Rather than running the combines `npm test`, and tried separate `npm run test:js` and `npm run test:types`, and it is only the latter that fails for me, while the former is running fine.

**Comment by RebeccaStevens** [comment]
Are you able to run `yarn build`?

**Comment by vitaly-t** [comment]
Yep, all issues look resolved 😄

**Comment by vitaly-t** [comment]
And then a little later, the issue with documents came back again, no idea why, as I didn't change anything...

![image](https://user-images.githubusercontent.com/5108906/194994629-8dc299c8-3ce1-47c5-8c4b-5616658048dc.png)

Any idea why this keeps happening?

**Comment by vitaly-t** [comment]
I researched it, and I was able to fix it by adding the following dependency:

```
"@types/estree": "1.0.0"
```

**Comment by vitaly-t** [comment]
@RebeccaStevens 

What's your current plan for the project? If we cannot make progress for the types, then maybe we should just focus on what's at hand - revise types and code for `flat` + `flatMap`, and make another Beta? :wink:

**Comment by RebeccaStevens** [comment]
I'll prioritize working on those with the current type system.

What were the breaking changes btw?

**Comment by vitaly-t** [comment]
> What were the breaking changes btw?

I didn't introduce any, there were only two new operators added.

You, however, mentioned breaking change through some type addition/change in #124. That was the whole reason I started 2.0.0 Beta versions. Please review those. And if you think we do not really have any breaking change, then there is no reason to up the version, and I can release 1.7.0 instead.

**Comment by RebeccaStevens** [comment]
Ahh, right. It's quite a simple change but yeah, it is breaking so we should do a major release.

**Comment by vitaly-t** [comment]
Great thanks! So let me add some docs here for the new operators + in general, and maybe you can look into #103 in the meantime, since you reported it? Then we can release 2.0.0, or at least I will do one more Beta first :wink:

**Comment by RebeccaStevens** [comment]
Will do.

#151 should fix the types for `flat`

**Comment by vitaly-t** [comment]
As I first [mentioned it in discussion](https://github.com/vitaly-t/iter-ops/discussions/131), here's what I think it should be:

1. `pipe` should remain as is, taking a mixed type of iterable, and handling mixed type casting.
2. `pipeSync` should throw an error when passed an `async` input, but only when called directly from the client (not from `pipe`)
3. `pipeAsync` should automatically convert to `async` when passed synchronous data

**Comment by vitaly-t** [comment]
To address what's outstanding, I have opened [this issue](https://github.com/vitaly-t/iter-ops/issues/177).

**Comment by RebeccaStevens** [comment]
I think for now we can mark this as done.
The majority of the happy-paths work.
To fix everything would probably require some fundamental restructuring of the library (if it's even possible).

**Comment by vitaly-t** [comment]
Do we really have that many problems? 🤦‍♂️ I thought you did full restructuring :smile:

What about that `flatMap` declaration? It doesn't look right to me, it's callback is supposed to allow promises, but it does not, while the implementation has it.

**Comment by RebeccaStevens** [comment]
The type issues should only show up when you go looking for them (such as sync operations working with async types (e.g. accepting promises)).

I'll look into that `flatMap` one.

**Comment by vitaly-t** [comment]
Great, I'll wait for this one before making another release :wink:

**Comment by vitaly-t** [comment]
Ok, I have published v2.3.0 with everything as is. This issue with `flatMap` can be addressed in the next update.
